faucetsdn/faucet
# Add Fault-Tolerance tests & generalise the Multi-DP integration tests

- Add Fault-tolerance tests. Fault-tolerance tests can be run in the same manner as the integration tests but with the environment variable FAUCET_TESTS="-t". Fault-Tolerance tests will build a topology then randomly (or statically) tear down links or switches.

```
sudo docker run --privileged --rm --net=host --cap-add=NET_ADMIN \
    -e FAUCET_TESTS="-t" \
    -ti faucet/tests
```

- Add FaucetTopoTestBase, provides a generalised form of the integration test FaucetStringOfDPTest. Remove FaucetStringOfDPTest, all previous integration tests that inherited from the FaucetStringOfDPTest now inherit from FaucetTopoTestBase. These tests have also been moved to a new file.

- Add FaucetTopoGenerator, provides a generalised form of FaucetStringOfDPSwitchTopo. Remove FaucetStringOfDPSwitchTopo. Use FaucetTopoGenerator instead.

- Add TopologyWatcher. The fault-tolerance tests use the TopologyWatcher to determine what hosts should be connected etc.

FaucetTopoTestBase and the FaucetTopoGenerator allows for any Networkx graph (these can be made with the [Networkx graph generators](https://networkx.github.io/documentation/stable/reference/generators.html)) to be turned into a Faucet topology. 

```
dp_links = FaucetTopoGenerator.dp_links_networkx_graph(networkx.cycle_graph(n_dps))
host_links, host_vlans = FaucetTopoGenerator.untagged_vlan_hosts(n_dps, n_vlans)
self.build_net(n_dps=n_dps, n_vlans=n_vlans, dp_links=dp_links,
                      host_links=host_links, host_vlans=host_vlans)
```
